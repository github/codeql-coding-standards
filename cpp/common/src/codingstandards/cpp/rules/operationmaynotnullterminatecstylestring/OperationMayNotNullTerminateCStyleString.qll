/**
 * Implements a shared library which detects operations that may not result in a
 * null-terminated c-style string.
 */

import cpp
import codingstandards.cpp.Customizations
import codingstandards.cpp.Exclusions
import semmle.code.cpp.models.implementations.StdString
import codingstandards.cpp.PossiblyUnsafeStringOperation

abstract class OperationMayNotNullTerminateCStyleStringSharedQuery extends Query { }

Query getQuery() { result instanceof OperationMayNotNullTerminateCStyleStringSharedQuery }

query predicate problems(PossiblyUnsafeStringOperation op, string message) {
  not isExcluded(op, getQuery()) and
  message = "Operation on a C-style string that may not result in a null-terminated string."
}
