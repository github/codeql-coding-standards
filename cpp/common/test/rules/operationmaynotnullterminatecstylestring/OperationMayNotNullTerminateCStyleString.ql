// GENERATED FILE - DO NOT MODIFY
import codingstandards.cpp.rules.operationmaynotnullterminatecstylestring.OperationMayNotNullTerminateCStyleString
