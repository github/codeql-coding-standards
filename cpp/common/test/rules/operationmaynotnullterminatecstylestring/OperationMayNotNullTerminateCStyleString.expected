| test.cpp:16:3:16:14 | call to strncpy | Operation on a C-style string that may not result in a null-terminated string. |
| test.cpp:17:3:17:14 | call to strncpy | Operation on a C-style string that may not result in a null-terminated string. |
| test.cpp:21:12:21:12 | call to operator>> | Operation on a C-style string that may not result in a null-terminated string. |
| test.cpp:27:6:27:9 | call to read | Operation on a C-style string that may not result in a null-terminated string. |
| test.cpp:40:3:40:15 | call to snprintf | Operation on a C-style string that may not result in a null-terminated string. |
| test.cpp:46:3:46:14 | call to strncat | Operation on a C-style string that may not result in a null-terminated string. |
